python-oauth-web-app/atproto_security.py

```diff
@@ -0,0 +1,41 @@
+from urllib.parse import urlparse
+import requests_hardened
+
+
+# this is a crude/partial filter that looks at HTTPS URLs and checks if they seem "safe" for server-side requests (SSRF). This is only a partial mitigation, the actual HTTP client also needs to prevent other attacks and behaviors.
+# this isn't a fully complete or secure implementation
+def is_safe_url(url):
+    parts = urlparse(url)
+    if not (
+        parts.scheme == "https"
+        and parts.hostname is not None
+        and parts.hostname == parts.netloc
+        and parts.username is None
+        and parts.password is None
+        and parts.port is None
+    ):
+        return False
+
+    segments = parts.hostname.split(".")
+    if not (
+        len(segments) >= 2
+        and segments[-1] not in ["local", "arpa", "internal", "localhost"]
+    ):
+        return False
+
+    if segments[0].isdigit():
+        return False
+
+    return True
+
+
+# configures a "hardened" requests wrapper
+hardened_http = requests_hardened.Manager(
+    requests_hardened.Config(
+        default_timeout=(2, 10),
+        never_redirect=True,
+        ip_filter_enable=True,
+        ip_filter_allow_loopback_ips=False,
+        user_agent_override="AtprotoCookbookOAuthFlaskDemo",
+    )
+)
```

Comment:
Is it possible to verify the url as part of `hardened_http` manager ? It will probably be less error prone than having to always have to call `is_safe_url` before making the request.

Reply:
It does some similar checks internally, but is an externally-maintained package. We also need to do the basic URL check for the authorization request URL before forwarding the browser; that code path doesn't involve the hardened client.

Reply:
Is it possible to have some `hardened_http_for_safe_url` version of the `hardened_http` that allows to replace:

```py
    assert is_safe_url(pds_url)
    with hardened_http.get_session() as sess:
```

with a single:

```py
    with hardened_http_for_safe_url.get_session() as sess:
```

I feel like it would be less error prone (but that may just be a nit from me)

Reply:
you are definitely correct, I just don't want to fork/maintain `requests_hardened` just for this example code